packages/composable-controller/src/ComposableController.ts

```diff
@@ -99,43 +107,108 @@ export function isBaseController(
+ */
+// TODO: Replace with `{ [name: string]: StateConstraint }` once BaseControllerV2 migrations are completed for all controllers.
+export type ComposableControllerStateConstraint = {
+  [name: string]: LegacyControllerStateConstraint;
 };
 
-export type ComposableControllerEvents = ComposableControllerStateChangeEvent;
-
-type AnyControllerStateChangeEvent = {
-  type: `${string}:stateChange`;
-  payload: [ControllerInstance['state'], Patch[]];
+/**
+ * A controller state change event for any controller instance that extends from either `BaseControllerV1` or `BaseControllerV2`.
+ */
+// TODO: Replace all instances with `ControllerStateChangeEvent` once `BaseControllerV2` migrations are completed for all controllers.
+type LegacyControllerStateChangeEvent<
+  ControllerName extends string,
+  ControllerState extends LegacyControllerStateConstraint,
+> = {
+  type: `${ControllerName}:stateChange`;
+  payload: [ControllerState, Patch[]];
 };
 
-type AllowedEvents = AnyControllerStateChangeEvent;
+export type ComposableControllerStateChangeEvent<
+  ComposableControllerState extends ComposableControllerStateConstraint,
+> = LegacyControllerStateChangeEvent<
+  typeof controllerName,
+  ComposableControllerState
+>;
+
+export type ComposableControllerEvents<
+  ComposableControllerState extends ComposableControllerStateConstraint,
+> = ComposableControllerStateChangeEvent<ComposableControllerState>;
+
+type ChildControllerStateChangeEvents<
+  ComposableControllerState extends ComposableControllerStateConstraint,
+> = ComposableControllerState extends Record<
+  infer ControllerName extends string,
+  infer ControllerState
+>
+  ? ControllerState extends StateConstraint
+    ? ControllerStateChangeEvent<ControllerName, ControllerState>
+    : ControllerState extends Record<string, unknown>
+    ? LegacyControllerStateChangeEvent<ControllerName, ControllerState>
+    : never
+  : never;
+
+type AllowedEvents<
+  ComposableControllerState extends ComposableControllerStateConstraint,
+> = ChildControllerStateChangeEvents<ComposableControllerState>;
 
-export type ComposableControllerMessenger = RestrictedControllerMessenger<
+export type ComposableControllerMessenger<
+  ComposableControllerState extends ComposableControllerStateConstraint,
+> = RestrictedControllerMessenger<
   typeof controllerName,
   never,
-  ComposableControllerEvents | AllowedEvents,
+  | ComposableControllerEvents<ComposableControllerState>
+  | AllowedEvents<ComposableControllerState>,
   never,
-  AllowedEvents['type']
+  AllowedEvents<ComposableControllerState>['type']
 >;
 
+type GetChildControllers<
+  ComposableControllerState,
+  P extends keyof ComposableControllerState = keyof ComposableControllerState,
+> = P extends string
+  ? ComposableControllerState[P] extends StateConstraint
+    ? { name: P; state: ComposableControllerState[P] }
+    : BaseControllerV1<
+        BaseConfig & Record<string, unknown>,
+        BaseState & ComposableControllerState[P]
```

Comment:
Nit: Can we use a non-one-letter name here for clarify? Perhaps `ControllerName`?

``` suggestion
  ControllerName extends keyof ComposableControllerState = keyof ComposableControllerState,
> = Key extends string
  ? ComposableControllerState[ControllerName] extends StateConstraint
    ? { name: ControllerName; state: ComposableControllerState[ControllerName] }
    : BaseControllerV1<
        BaseConfig & Record<string, unknown>,
        BaseState & ComposableControllerState[ControllerName]
```

Reply:
Fixed here: https://github.com/MetaMask/core/pull/3952/commits/25f2f0b6574dc5dda7cc7832cbc5a8833f240e75